File: gamerental/src/test/java/com/example/app/user/UserTest.java
package com.example.app.user;

import io.fluxzero.sdk.test.TestFixture;
import io.fluxzero.sdk.tracking.handling.authentication.UnauthorizedException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.List;

class UserTest {

    final TestFixture testFixture = TestFixture.create();

    @Test
    void createViewer() {
        testFixture.whenCommand("/user/create-customer.json")
                .expectEvents("/user/create-customer.json");
    }


    @Test
    void createUserAllowedAsEditor() {
        testFixture
                .givenCommands("/user/create-admin.json")
                .whenCommandByUser("admin", "/user/create-customer.json")
                .expectEvents("/user/create-customer.json");
    }

    @Test
    void createUserNotAllowedAsViewer() {
        testFixture.whenCommandByUser("viewer", "/user/create-admin.json")
                .expectExceptionalResult(UnauthorizedException.class);
    }

    @Test
    void getUsers() {
        testFixture.givenCommands("/user/create-customer.json")
                .whenQuery(new GetUsers())
                .expectResult(r -> r.size() == 1);
    }

    @Nested
    class UsersEndpointTests {

        @BeforeEach
        void setUp() {
            testFixture.registerHandlers(new UsersEndpoint());
        }

        @Test
        void createUser() {
            testFixture.whenPost("/users", "/user/new-user-details.json")
                    .expectResult(UserId.class)
                    .expectEvents(CreateUser.class);
        }

        @Test
        void getUsers() {
            testFixture.givenPost("/users", "/user/new-user-details.json")
                    .whenGet("/users")
                    .<List<UserProfile>>expectResult(r -> r.size() == 1);
        }
    }
}

File: flux-basic-kotlin/src/test/kotlin/com/example/app/user/UserTest.kt
package com.example.app.user

import com.example.app.user.api.GetUsers
import com.example.app.user.api.model.UserProfile
import io.fluxzero.sdk.test.TestFixture
import io.fluxzero.sdk.tracking.handling.IllegalCommandException
import io.fluxzero.sdk.tracking.handling.authentication.UnauthorizedException
import org.junit.jupiter.api.Test

class UserTest {

    private val testFixture = TestFixture.create()

    @Test
    fun createViewer() {
        testFixture.whenCommand("/user/create-user.json")
            .expectEvents("/user/create-user.json")
    }

    @Test
    fun createUser() {
        testFixture
            .whenCommand("/user/create-user.json")
            .expectEvents("/user/create-user.json")
    }

    @Test
    fun createUserTwiceForbidden() {
        testFixture
            .givenCommands("/user/create-user.json")
            .whenCommand("/user/create-user.json")
            .expectExceptionalResult(IllegalCommandException::class)
    }

    @Test
    fun createUserNotAllowedForNonAdmin() {
        testFixture.whenCommandByUser("viewer", "/user/create-admin.json")
            .expectExceptionalResult(UnauthorizedException::class.java)
    }

    @Test
    fun setRole() {
        testFixture
            .givenCommands("/user/create-user.json")
            .whenCommand("/user/make-admin.json")
            .expectEvents("/user/make-admin.json")
    }

    @Test
    fun setRoleAsNonAdminNotAllowed() {
        testFixture
            .givenCommands("/user/create-user.json")
            .whenCommandByUser("viewer", "/user/make-admin.json")
            .expectExceptionalResult(UnauthorizedException::class.java)
    }

    @Test
    fun getUsers() {
        testFixture.givenCommands("/user/create-user.json")
            .whenQuery(GetUsers())
            .expectResult<List<UserProfile>> { r: List<UserProfile> -> r.size == 1 }
    }

}
